From 53246ba7b026c92e42aa4c332007f425d297a4ea Mon Sep 17 00:00:00 2001
From: Ali Mohammad Pur <ali.mpfard@gmail.com>
Date: Mon, 16 May 2022 16:21:08 +0430
Subject: [PATCH 3/3] Make jv_mem_alloc() return a dummy allocation for
 zero-sized allocs

And make jv_mem_free() accept it.
---
 src/jv_alloc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/jv_alloc.c b/src/jv_alloc.c
index fd7b257..9f5ced8 100644
--- a/src/jv_alloc.c
+++ b/src/jv_alloc.c
@@ -130,7 +130,14 @@ void* jv_mem_alloc_unguarded(size_t sz) {
   return malloc(sz);
 }
 
+static void* jv_mem_calloc_unique_allocd = 0;
 void* jv_mem_calloc(size_t nemb, size_t sz) {
+  if (!sz || !nemb) {
+      if (jv_mem_calloc_unique_allocd)
+          return jv_mem_calloc_unique_allocd;
+      else
+          return jv_mem_calloc_unique_allocd = calloc(1,1);
+  }
   void* p = calloc(nemb, sz);
   if (!p) {
     memory_exhausted();
@@ -155,6 +162,8 @@ char* jv_mem_strdup_unguarded(const char *s) {
 }
 
 void jv_mem_free(void* p) {
+  if (p == jv_mem_calloc_unique_allocd)
+    return;
   free(p);
 }
 
-- 
2.36.1

